terraform/eval_diff_test.go

```diff
@@ -0,0 +1,78 @@
+package terraform
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestEvalFilterDiff(t *testing.T) {
+	ctx := new(MockEvalContext)
+
+	cases := []struct {
+		Node   *EvalFilterDiff
+		Input  *InstanceDiff
+		Output *InstanceDiff
+	}{
+		// With no settings, it returns an empty diff
+		{
+			&EvalFilterDiff{},
+			&InstanceDiff{Destroy: true},
+			&InstanceDiff{},
+		},
+
+		// Destroy
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{Destroy: false},
+			&InstanceDiff{Destroy: false},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{Destroy: true},
+			&InstanceDiff{Destroy: true},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{
+				Destroy: true,
+				Attributes: map[string]*ResourceAttrDiff{
+					"foo": &ResourceAttrDiff{},
+				},
+			},
+			&InstanceDiff{Destroy: true},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{
+				Attributes: map[string]*ResourceAttrDiff{
+					"foo": &ResourceAttrDiff{
+						RequiresNew: true,
+					},
+				},
+			},
+			&InstanceDiff{Destroy: true},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{
+				Attributes: map[string]*ResourceAttrDiff{
+					"foo": &ResourceAttrDiff{},
+				},
+			},
+			&InstanceDiff{Destroy: false},
+		},
+	}
+
+	for i, tc := range cases {
+		var output *InstanceDiff
+		tc.Node.Diff = &tc.Input
+		tc.Node.Output = &output
+		if _, err := tc.Node.Eval(ctx); err != nil {
+			t.Fatalf("err: %s", err)
+		}
+
+		if !reflect.DeepEqual(output, tc.Output) {
+			t.Fatalf("bad: %d\n\n%#v", i, output)
```

Comment:
@mitchellh We've talked about this a bit, and I know you were flying forward on a critical bug here so I'm not talking about this specific instance, but for new test case harnesses can we agree to prefer:

 * Maps with test names instead of integer indexes
 * Failure output that includes test name, expected output, and actual output (instead of just "bad: actual")

Perhaps I am a dummy, but for tests written in this style it takes me a lot of mental ramp time to be able to jump in and read these coherently. My theory is that a few tweaks today can provide compounding savings of future developer time.